docs/decisions/00NN-text-search.md

```diff
@@ -0,0 +1,986 @@
+
+For more information and the latest updates from the Semantic Kernel team, you can visit their [official blog](https://devblogs.microsoft.com/semantic-kernel/).
+```
+
+In the previous example the site has hard coded. It is also possible to allow the LLM to extract the site from the user query. In the example below a custom search function is created which includes an additional argument to allow the LLM to set the site.
+
+```csharp
+// Create a kernel with OpenAI chat completion
+IKernelBuilder kernelBuilder = Kernel.CreateBuilder();
+kernelBuilder.AddOpenAIChatCompletion(
+        modelId: TestConfiguration.OpenAI.ChatModelId,
+        apiKey: TestConfiguration.OpenAI.ApiKey,
+        httpClient: httpClient);
+Kernel kernel = kernelBuilder.Build();
+
+// Create a search service with Bing search service
+var textSearch = new BingTextSearch(new(TestConfiguration.Bing.ApiKey));
+
+// Build a text search plugin with Bing search service and add to the kernel
+var searchPlugin = KernelPluginFactory.CreateFromFunctions("SearchPlugin", "Search Microsoft Dev Blogs site", [CreateSearchBySite(textSearch)]);
+kernel.Plugins.Add(searchPlugin);
+
+// Invoke prompt and use text search plugin to provide grounding information
+OpenAIPromptExecutionSettings settings = new() { ToolCallBehavior = ToolCallBehavior.AutoInvokeKernelFunctions };
+KernelArguments arguments = new(settings);
+Console.WriteLine(await kernel.InvokePromptAsync("What is the Semantic Kernel? Only include results from devblogs.microsoft.com. Include citations to the relevant information where it is referenced in the response.", arguments));
+```
+
+The code below shows how the custom search function is created.
+
+- The `KernelFunction` includes an additional optional parameter called `site`
+- If the `site` parameter is provided a `BasicFilterOptions` instance is created which will cause Bing to return responses only from that site
+- A custom function description and parameter description are provided to help the LLM in using this method.
+
+```csharp
+private static KernelFunction CreateSearchBySite(ITextSearch<BingWebPage> textSearch, BasicFilterOptions? basicFilter = null, MapSearchResultToString? mapToString = null)
+{
+    async Task<IEnumerable<BingWebPage>> SearchAsync(Kernel kernel, KernelFunction function, KernelArguments arguments, CancellationToken cancellationToken)
+    {
+        arguments.TryGetValue("query", out var query);
+        if (string.IsNullOrEmpty(query?.ToString()))
+        {
+            return [];
+        }
+
+        var parameters = function.Metadata.Parameters;
+
+        arguments.TryGetValue("count", out var count);
+        arguments.TryGetValue("skip", out var skip);
+        arguments.TryGetValue("site", out var site);
+        BasicFilterOptions? basicFilter = null;
+        if (string.IsNullOrEmpty(site?.ToString()))
+        {
+            basicFilter = new BasicFilterOptions().Equality("site", site?.ToString()!);
+        }
```

Comment:
Once we get to implementation, it'll be nice to have a built in variation of this whole method, that just takes a function name, description and kernel parameter metadata in addition to query count and skip.
That way, someone can just pass in the function name, description and a parameter for site, and here in the highlighted section, we can loop over the passed in parameters, and just get them one by one from arguments, and add them to the basic filter automatically.

That way, someone doesn't have to write any of this for the simple case where they just want another parameter that should be passed along as a filter option.

Reply:
Agreed, great suggestion. Created a task to track this: https://github.com/microsoft/semantic-kernel/issues/8328